applications/crossbar/src/modules/cb_migrations.erl

```diff
@@ -0,0 +1,273 @@
+    cb_context:set_resp_data(Context, Result).
+
+-spec maybe_filter_resellers(list(), cb_context:context()) -> list(kz_json:object()).
+maybe_filter_resellers(Accounts, Context) ->
+    case kz_json:get_binary_boolean(<<"include_resellers">>, cb_context:req_data(Context), <<"false">>) of
+        <<"true">>  -> Accounts;
+        <<"false">> -> filter_account_resellers(Accounts)
+    end.
+
+-spec filter_account_descendants(binary(), list(), list()) -> list().
+filter_account_descendants(Account, [H|T], Acc) ->
+    case kz_json:get_value(<<"key">>, H) of
+        [Account, _] -> filter_account_descendants(Account, T, [H|Acc]);
+        _OtherValue  -> filter_account_descendants(Account, T, Acc)
+    end;
+
+filter_account_descendants(_Account, [], Acc) ->
+    Acc.
+
+-spec filter_account_resellers(list()) -> list().
+filter_account_resellers(Accounts) ->
+    Resellers = get_resellers(Accounts, []),
+    filter_account_resellers(Accounts, Resellers, []).
+
+-spec filter_account_resellers(list(), list(), list()) -> list().
+filter_account_resellers([H|T], Res, Acc) ->
+    Tree = kz_json:get_value([<<"value">>, <<"tree">>], H),
+
+    case lists:subtract(Tree, Res) of
+        New when length(New) < length(Tree) ->
+            filter_account_resellers(T, Res, Acc);
+
+        _NotInList ->
+            filter_account_resellers(T, Res, [H|Acc])
+    end;
+
+filter_account_resellers([], _Res, Acc) ->
+    Acc.
+
+-spec get_resellers(list(), list()) -> list().
+get_resellers([H|T], Acc) ->
+    Account = kz_json:get_value(<<"id">>, H),
+    case kz_services:is_reseller(Account) of
+        'true'  -> get_resellers(T, [Account|Acc]);
+        'false' -> get_resellers(T, Acc)
+    end;
+
+get_resellers([], Acc) ->
+    Acc.
+
+-spec migrate_on_account(binary(), binary(), atom(), cb_context:context()) -> binary().
+migrate_on_account(Account, MigId, Module, Context) ->
+    Result = Module:perform_migration(Account, Context),
+    maybe_mark_as_complete(MigId, Account, Result),
+    Account.
+
+-spec maybe_mark_as_complete(binary(), binary(), cb_context:context()) -> 'ok'.
+maybe_mark_as_complete(MigId, Account, Context) ->
+    case cb_context:resp_status(Context) of
+        'success' ->
+            mark_migration_complete(MigId, Account, Context);
+
+        _Failed ->
+            lager:info("migration ~p failed to complete", [MigId])
+    end.
+
+-spec mark_migration_complete(binary(), binary(), cb_context:context()) -> 'ok'.
+mark_migration_complete(MigId, AccountId, Context) ->
+    lager:info("migration ~p completed successfully on account ~p", [MigId, AccountId]),
+
+    AccountDb = kz_util:format_account_id(AccountId, 'encoded'),
+    maybe_create_migration_doc(AccountDb),
+
+    {'ok', Doc} = kz_datamgr:open_cache_doc(AccountDb, ?MIGRATIONS_DOC),
+    Migrations  = kz_json:get_value(<<"migrations_performed">>, Doc),
+
+    Args = kz_json:from_list([
+                              {<<"authorizing_id">>, cb_context:auth_user_id(Context)}
```

Comment:
@mark2600 it would be good to capture the auth account id, auth account name and auth user first/last name as well.